File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int left = 0, right = 0;

        for (int w : weights) {
            left = Math.max(left, w);
            right += w;
        }

        while (left < right) {
            int mid = left + (right - left) / 2;

            int d = 1, sum = 0;

            for (int w : weights) {
                if (sum + w > mid) {
                    d++;
                    sum = 0;
                }
                sum += w;
            }

            if (d <= days)
                right = mid;
            else
                left = mid + 1;
        }

        return left;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights, days):
        l, r = max(weights), sum(weights)

        while l < r:
            mid = (l + r) // 2
            d, total = 1, 0

            for w in weights:
                if total + w > mid:
                    d += 1
                    total = 0
                total += w

            if d <= days:
                r = mid
            else:
                l = mid + 1

        return l